### DISCUSSION TITLE
Channel creation using fabric-admin-sdk

### DISCUSSION THREAD

**Discussion by adityajoshi12** [hyperledger/fabric-admin-sdk#187] [Q&A]
I wanted to use this sdk in my application to be able to create channels on the runtime and I was going through the testcases to understand how to use this sdk. I found that in the testcase we are using packages that are internal to this project, which makes it difficult to use this sdk in the application.

> https://github.com/hyperledger/fabric-admin-sdk/blob/d480efd5dca268ff82ffa0db3fa6799b362682bb/test/e2e_test.go#L17

Also will there be any high level API that support the creation of configtx programatically. This will remove the dependency of the `Configtx.yaml` and channel blocks can be generated using the sdk.

**Comment by SamYuan1990** [comment]
> An internal to this project, which makes it difficult to use this sdk in the application.

@adityajoshi12 could you please more specific about why the internal package makes difficult in your case?

> remove the dependency of the `Configtx.yaml`?

if we don't have a config tx yaml... there seems lots of input values into the function for generate the block right? I am confused a little. hence need a more specific/clear description.

**Reply by adityajoshi12** [reply]
I am trying to create channel using the sdk in my application, now in order to create/access the profile `profile, err = genesisconfig.Load("ChannelUsingRaft", "./")` I need to access the `genesisconfig` package which i can't access since it's a internal package.

Other part: if there would be some sort of function that create the genesis block required to create channel. This kind of feature is available in legecy fabric-sdk-java [here](https://github.com/hyperledger/fabric-sdk-java/blob/main/src/main/java/org/hyperledger/fabric/sdk/HFClient.java#L273). This will allow channel creation without the need of configtx file and make it indepenent.

**Reply by SamYuan1990** [reply]
+ @bestbeforetoday , @denyeart in loop
currently the admin-sdk try to keep same behavior with cli which read a config profile for block generation. IMO from code structure point of view seems no harmful to make the packages from internal to publish. 
wdyt?

**Comment by adityajoshi12** [comment]
Also, this align with the initial idea of @denyeart mentioned [here](https://github.com/hyperledger/fabric-admin-sdk/issues/15#issuecomment-1167832370)

**Reply by SamYuan1990** [reply]
let's wait for @denyeart 's feedback. 
I suppose @adityajoshi12 your point is that we just make the channel config package from internal to public for fabric admin sdk repo and replace with https://github.com/hyperledger/fabric-config ?

**Reply by adityajoshi12** [reply]
I am looking for two things:
1. Moving the `configtxgen` out of internal pkg. This can be immediate fix.
2. The other ask would be to have a high level API that allows the channel creation without the need of `configtx.yaml` file.

**Comment by adityajoshi12** [comment]
Hi @SamYuan1990 , it been a long, we didn't had any update on this.

**Reply by SamYuan1990** [reply]
I am waiting for other maintains feedback as well. @denyeart and @bestbeforetoday what do you think?

**Reply by SamYuan1990** [reply]
I listed this discussion on fabric contributor meeting. https://wiki.hyperledger.org/display/fabric/Contributor+Meetings+2024

**Reply by SamYuan1990** [reply]
@adityajoshi12 , here is feedback from contributor meeting.
https://github.com/hyperledger/fabric-admin-sdk/discussions/187#discussioncomment-8877618 the sample here as you mentioned is node sdk but the link is java and not one for channel creation.
So, we want to request more details.

at the meanwhile, as fabric-config project, if publish the package really benefits people, we are happy to make it. Hence I am going to make a PR, but we should evaluate if it's good enough to public.

**Reply by adityajoshi12** [reply]
I have updated the link above. The requirement here is to create a channel programatically using the admin-sdk. Currently admin-sdk has capability to create the channel using the `Configtx.yaml` file, but I would be interested in leveraging admin-sdk create channel using the high level API, that will avoid us from the need of `Configtx.yaml` file.

The other requirement is to make the `configtxgen` pkg out of internal pkg so that functions like `genesisconfig.Load("ChannelUsingRaft", "./")` can be accessed outside in client applications.

**Reply by SamYuan1990** [reply]
Pr opened as https://github.com/hyperledger/fabric-admin-sdk/pull/193
